src/order_watcher.ts

```diff
@@ -66,8 +66,11 @@ export class OrderWatcher implements OrderWatcherInterface {
         if (filledOrders.length > 0) {
             throw new Error(`already fully filled orders ${JSON.stringify(filledOrders)}`);
         }
-        // Saves all given entities in the database. If entities do not exist in the database then inserts, otherwise updates.
-        await this._connection.getRepository(SignedOrderV4Entity).save(validOrders);
+        // Saves all specified entities with hash duplicates removed in the database. If entities do not exist in the database then inserts, otherwise updates.
+        const uniqueValidOrders = Array.from(
+            new Map(validOrders.map((order) => [order.hash, order]))
+        );
```

Comment:
この問題をさらに調査するとtypeorm自体にissueが作成されていました。
https://github.com/typeorm/typeorm/issues/4122
このissueを読んでみたところsave処理の前に重複を取り除くことは提供されているworkaroundの1つのようです。
https://github.com/typeorm/typeorm/issues/4122#issuecomment-779440998
このworkaroundでもうまくいかなかった場合は以下の方法を試す必要があるかもしれません。
https://github.com/typeorm/typeorm/issues/4122#issuecomment-569924481

Reply:
> question: typeormのsave()って既に同じデータがDBにある場合、エラー出ますか？

https://orkhan.gitbook.io/typeorm/docs/repository-api
```
save - Saves a given entity or array of entities. If the entity already exist in the database, it is updated. If the entity does not exist in the database, it is inserted. It saves all given entities in a single transaction (in the case of entity, manager is not transactional). Also supports partial updating since all undefined properties are skipped. Returns the saved entity/entities.
```

ドキュメントを読んでみると、同じデータが存在する場合は更新処理をするとのことです。そのためエラーは発生しないと考えています。

複数のentityをsaveで処理したい時にこのエラーが発生しているのではないかと考えています。